### PR TITLE
RTL6a

### PR DESCRIPTION
**Pull request by ricardopereira**

### CODE REVIEW

**Comment by ricardopereira** [comment]
I'm getting errors after the #171:

```
2016-02-04 18:50:15.161 xctest[85089:13705808] ERROR: ARTDataDecoded failed to decode data as 'json': (__NSCFDictionary){
    value = 1;
}
2016-02-04 18:50:15.162 xctest[85089:13705808] ERROR: ARTRealtimeChannel: error decoding data, status: 2038569904
```

@tcard Any thoughts?

**Comment by ricardopereira** [comment]
I can debug it later.

**Comment by tcard** [comment]
Seems like it's trying to decode a piece of JSON that is already decoded. The `data` of the incoming message should be a string, not a dictionary

**Comment by ricardopereira** [comment]
The test is publishing a dictionary. The existing `RSL4d4` test is also failing.
@tcard Should I open an issue?

**Comment by tcard** [comment]
@ricardopereira Please.

File: ablySpec/RealtimeClientChannel.swift [hunk 1]
```diff
@@ -186,6 +187,56 @@ class RealtimeClientChannel: QuickSpec {
...
+                    let data = ["value":1]
+                    var restEncodedMessage: ARTMessage?
+                    var realtimeEncodedMessage: ARTMessage?
+
+                    let rest = ARTRest(options: AblyTests.commonAppSetup())
+                    let restChannel = rest.channels.get("test")
+
+                    let restBlock: @convention(block) (AspectInfo, ARTMessage) -> Void = { info, message in
+                        let invocation = info.originalInvocation()
+                        invocation.invoke()
```
Comment (tcard) [inline ablySpec/RealtimeClientChannel.swift]: I'm not sure how this aspects thing works, but if `message` here is the original, unencoded message that is passed to `encodeMessageIfNeeded`, take into account that `ARTMessage.encode` returns a _copy_ of the ARTMessage instance, and thus this `message` won't be mutated by `invocation.invoke()`, and you're assigning the unencoded version to `restEncodedMessage`.

Ignore this if this `message` is what's returned by `encodeMessageIfNeeded`. (But then I don't understand what `invocation.invoke()` is invoking.)

Comment (ricardopereira) [inline ablySpec/RealtimeClientChannel.swift]: `invocation.invoke()` calls the original implementation, so the `encodeWithEncoder` will write to the current `ARTMessage` instance (look https://github.com/ably/ably-ios/blob/bff8f901b67e9ab12b38fa477470ed773ae0a902/ably-ios/ARTChannel.m#L54). Even so, I saw the documentation and I think it is possible to retrieve the original return value of the mehotd directly with the [NSInvocation.getReturnValue](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSInvocation_Class/#//apple_ref/occ/instm/NSInvocation/getReturnValue:). I think both approaches are correct. What you think?

Comment (tcard) [inline ablySpec/RealtimeClientChannel.swift]: That link shows how `encodeWithEncoder` takes an `output` argument, where it writes the _new_ message, which is then a different instance than the `message` in this block. For this to be correct you need to use the returned message.

Comment (ricardopereira) [inline ablySpec/RealtimeClientChannel.swift]: Yes, you right, It is doing a copy of the `message`. I will try to use the `NSInvocation.getReturnValue`. Ty.

File: ablySpec/RealtimeClientChannel.swift [hunk 2]
```diff
@@ -186,6 +187,56 @@ class RealtimeClientChannel: QuickSpec {
...
+                    let _ = try? realtimeHook("encodeMessageIfNeeded:", withOptions: .PositionAfter, usingBlock: unsafeBitCast(realtimeBlock, ARTRealtimeChannel.self))
+
+                    waitUntil(timeout: testTimeout) { done in
+                        realtimeChannel.publish(data, cb: { status in
+                            expect(status.state).to(equal(ARTState.Ok))
+                            done()
+                        })
+                    }
+
+                    expect(restEncodedMessage!.data as? NSObject).to(equal(realtimeEncodedMessage!.data as? NSObject))
```
Comment (mattheworiordan) [inline ablySpec/RealtimeClientChannel.swift]: Is it worth checking that the encoding value matches as well, and that encoding is not null i.e. the messages are indeed encoded.

File: ablySpec/RealtimeClientChannel.swift (hunk 1 above)
Comment (ricardopereira) [inline ablySpec/RealtimeClientChannel.swift]: Fixed 357c5db.

File: ablySpec/RealtimeClientChannel.swift (hunk 2 above)
Comment (ricardopereira) [inline ablySpec/RealtimeClientChannel.swift]: Done 357c5db.

**Comment by mattheworiordan** [comment]
Ok, looks good to me

**Comment by tcard** [comment]
LGTM

**Pull request merged**